liquidity_pool/src/liquidity.rs

```diff
@@ -217,39 +215,55 @@ pub trait LiquidityModule:
         );
         let mut borrow_position = self.borrow_position(borrow_token_nonce).get();
 
-        let accumulated_debt =
-            self.get_debt_interest(borrow_token_amount, borrow_position.timestamp)?;
-        let total_owed = borrow_token_amount + &accumulated_debt;
+        let accumulated_debt = self.get_debt_interest(asset_amount, borrow_position.timestamp)?;
+        let total_owed = asset_amount + &accumulated_debt;
 
         require!(
             asset_amount >= &total_owed,
             "Not enough asset tokens deposited"
         );
 
-        let extra_asset_paid = asset_amount - &total_owed;
-        if extra_asset_paid > 0 {
+        if asset_amount > &total_owed {
+            let extra_asset_paid = asset_amount - &total_owed;
             self.send()
                 .direct(&initial_caller, asset_token_id, 0, &extra_asset_paid, &[]);
         }
 
-        // TODO: Instead of borrow_token_amount (i.e. 1:1 ratio), calculate how much collateral amount was repaid
-        borrow_position.lend_tokens.amount -= borrow_token_amount;
-        if borrow_position.lend_tokens.amount == 0 {
+        if self.is_full_repay(&borrow_position, borrow_token_amount) {
+            lend_token_amount_to_send_back = borrow_position.lend_tokens.amount;
             self.borrow_position(borrow_token_nonce).clear();
         } else {
+            lend_token_amount_to_send_back = self.rule_of_three(
+                &borrow_position.lend_tokens.amount,
+                borrow_token_amount,
+                &borrow_position.borrowed_amount,
+            );
+
+            require!(
+                lend_token_amount_to_send_back > 0,
+                "repay too little. lend tokens amount is zero"
+            );
+
+            borrow_position.borrowed_amount -= borrow_token_amount;
+            borrow_position.lend_tokens.amount -= &lend_token_amount_to_send_back;
             self.borrow_position(borrow_token_nonce)
                 .set(&borrow_position);
         }
 
+        self.borrowed_amount()
+            .update(|total| *total -= borrow_token_amount);
```

Comment:
I don't think this is right. The `borrowed_amount` keeps the amount of asset tokens borrowed by the user. When the user borrows, they receive X "borrow tokens", one for each token used as payment for borrow, and Y of the asset tokens.

In the borrow endpoint, `borrowed_amount` is incremeted with the amount of asset the user received, not by the amount they deposited as payment, and as such, this should be decremented by the amount of asset repaid, without the debt included.

Either way, this `borrowed_amount` concept is flawed, as it can actually lead to negative values. Say I borrow with 10 ETH collateral, at a ratio of 1 ETH = $200, I receive, let's say $1800 of the $2000 value in stablecoins.

borrowed_amount = 1800

I come back later, when the price is 1 ETH = $250. Assume some random value like $4000 as total with debt included. The repay endpoint will try to do `1800 - 4000` and it will fail.

Will have to rethink about how we keep this `borrowed_amount` and how we map it in case of price changes. 

To fix this: I think we should store that $200 price and the borrow rate in the borrow metadata, and at repay time, we see the user repaid 10 ETH, so we re-calculate that $1800 amount and update `borrowed_amount` accordingly.

Sorry for wall of text :sweat_smile: